File: app/schemas/data_dictionary.py
"""
Schemas for Regulatory Data Dictionary
"""

from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime


class RegulatoryDataDictionaryBase(BaseModel):
    """Base schema for regulatory data dictionary"""
    report_name: str = Field(..., description="Report name")
    schedule_name: str = Field(..., description="Schedule name")
    line_item_number: Optional[str] = Field(None, description="Line item number")
    line_item_name: str = Field(..., description="Line item name")
    technical_line_item_name: Optional[str] = Field(None, description="Technical line item name")
    mdrm: Optional[str] = Field(None, description="Master Data Reference Model")
    description: Optional[str] = Field(None, description="Description")
    static_or_dynamic: Optional[str] = Field(None, description="Static or Dynamic")
    mandatory_or_optional: Optional[str] = Field(None, description="Mandatory or Optional")
    format_specification: Optional[str] = Field(None, description="Format specification")
    num_reports_schedules_used: Optional[str] = Field(None, description="Number of reports/schedules used")
    other_schedule_reference: Optional[str] = Field(None, description="Other schedule reference")
    is_active: bool = Field(default=True, description="Is active")


class RegulatoryDataDictionaryCreate(RegulatoryDataDictionaryBase):
    """Schema for creating regulatory data dictionary entry"""
    pass


class RegulatoryDataDictionaryUpdate(BaseModel):
    """Schema for updating regulatory data dictionary entry"""
    report_name: Optional[str] = Field(None, description="Report name")
    schedule_name: Optional[str] = Field(None, description="Schedule name")
    line_item_number: Optional[str] = Field(None, description="Line item number")
    line_item_name: Optional[str] = Field(None, description="Line item name")
    technical_line_item_name: Optional[str] = Field(None, description="Technical line item name")
    mdrm: Optional[str] = Field(None, description="Master Data Reference Model")
    description: Optional[str] = Field(None, description="Description")
    static_or_dynamic: Optional[str] = Field(None, description="Static or Dynamic")
    mandatory_or_optional: Optional[str] = Field(None, description="Mandatory or Optional")
    format_specification: Optional[str] = Field(None, description="Format specification")
    num_reports_schedules_used: Optional[str] = Field(None, description="Number of reports/schedules used")
    other_schedule_reference: Optional[str] = Field(None, description="Other schedule reference")
    is_active: Optional[bool] = Field(None, description="Is active")


class RegulatoryDataDictionaryResponse(RegulatoryDataDictionaryBase):
    """Schema for regulatory data dictionary response"""
    dict_id: int = Field(..., description="Dictionary ID")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    
    model_config = ConfigDict(from_attributes=True)


class RegulatoryDataDictionaryListResponse(BaseModel):
    """Schema for regulatory data dictionary list response"""
    items: List[RegulatoryDataDictionaryResponse] = Field(..., description="List of dictionary entries")
    total: int = Field(..., description="Total number of entries")
    page: int = Field(..., description="Current page")
    per_page: int = Field(..., description="Items per page")


class DataDictionaryFilter(BaseModel):
    """Schema for filtering data dictionary"""
    report_name: Optional[str] = Field(None, description="Filter by report name")
    schedule_name: Optional[str] = Field(None, description="Filter by schedule name")
    line_item_name: Optional[str] = Field(None, description="Filter by line item name (contains)")
    mandatory_or_optional: Optional[str] = Field(None, description="Filter by mandatory/optional")
    static_or_dynamic: Optional[str] = Field(None, description="Filter by static/dynamic")
    search: Optional[str] = Field(None, description="General search term")
    is_active: Optional[bool] = Field(None, description="Filter by active status")


class DataDictionaryImportRequest(BaseModel):
    """Schema for importing data dictionary entries to report attributes"""
    selected_dict_ids: List[int] = Field(..., description="List of dictionary IDs to import")
    cycle_id: int = Field(..., description="Target cycle ID")
    report_id: int = Field(..., description="Target report ID")
    import_options: Optional[dict] = Field(None, description="Import options (e.g., override existing)")


class DataDictionaryImportResponse(BaseModel):
    """Schema for data dictionary import response"""
    success: bool = Field(..., description="Import success status")
    imported_count: int = Field(..., description="Number of entries imported")
    skipped_count: int = Field(..., description="Number of entries skipped")
    error_count: int = Field(..., description="Number of entries with errors")
    messages: List[str] = Field(default=[], description="Import messages")
    created_attributes: List[int] = Field(default=[], description="IDs of created attributes") 

File: frontend/src/api/dataDictionary.ts
import apiClient from './client';

export interface RegulatoryDataDictionaryEntry {
  dict_id: number;
  report_name: string;
  schedule_name: string;
  line_item_number?: string;
  line_item_name: string;
  technical_line_item_name?: string;
  mdrm?: string;
  description?: string;
  static_or_dynamic?: string;
  mandatory_or_optional?: string;
  format_specification?: string;
  num_reports_schedules_used?: string;
  other_schedule_reference?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DataDictionaryListResponse {
  items: RegulatoryDataDictionaryEntry[];
  total: number;
  page: number;
  per_page: number;
}

export interface DataDictionaryFilter {
  report_name?: string;
  schedule_name?: string;
  line_item_name?: string;
  mandatory_or_optional?: string;
  static_or_dynamic?: string;
  search?: string;
  is_active?: boolean;
}

export interface DataDictionaryImportRequest {
  selected_dict_ids: number[];
  cycle_id: number;
  report_id: number;
  import_options?: Record<string, any>;
}

export interface DataDictionaryImportResponse {
  success: boolean;
  imported_count: number;
  skipped_count: number;
  error_count: number;
  messages: string[];
  created_attributes: number[];
}

export interface DataDictionaryStats {
  total_entries: number;
  total_reports: number;
  total_schedules: number;
  mandatory_count: number;
  optional_count: number;
}

class DataDictionaryAPI {
  async getDataDictionary(
    page: number = 1,
    perPage: number = 50,
    filters?: DataDictionaryFilter
  ): Promise<DataDictionaryListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      per_page: perPage.toString(),
    });

    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          // Don't use params.append for values with special characters
          // The URLSearchParams will handle encoding automatically
          params.set(key, value.toString());
        }
      });
    }

    const response = await apiClient.get(`/data-dictionary/?${params.toString()}`);
    return response.data;
  }

  async getAvailableReports(): Promise<string[]> {
    const response = await apiClient.get('/data-dictionary/reports/');
    return response.data;
  }

  async getAvailableSchedules(reportName?: string): Promise<string[]> {
    const params = reportName ? `?report_name=${encodeURIComponent(reportName)}` : '';
    const response = await apiClient.get(`/data-dictionary/schedules${params}`);
    return response.data;
  }

  async importDataDictionaryEntries(
    importRequest: DataDictionaryImportRequest
  ): Promise<DataDictionaryImportResponse> {
    const response = await apiClient.post('/data-dictionary/import/', importRequest);
    return response.data;
  }

  async reloadDataDictionary(clearExisting: boolean = false): Promise<any> {
    const params = clearExisting ? '?clear_existing=true' : '';
    const response = await apiClient.post(`/data-dictionary/reload${params}`);
    return response.data;
  }

  async getDataDictionaryStats(): Promise<DataDictionaryStats> {
    const response = await apiClient.get('/data-dictionary/stats/');
    return response.data;
  }
}

export const dataDictionaryAPI = new DataDictionaryAPI(); 